### PR TITLE
views: addition of elasticsearch error handler

### LINKED ISSUE #84: views: handle elasticsearch errors
If you use `invenio_records_rest.query.es_search_factory` and the search phrase is invalid `elasticsearch_py` will raise an error which will result in a HTTP 500 error. Instead we should handle the error and reply with HTTP 400, and a proper error message explaining that the query was invalid.

Below is a stack trace of the error:

``` trackback
Stacktrace (most recent call last):

  File "flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "flask_cors/extension.py", line 188, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "invenio_rest/views.py", line 206, in dispatch_request
    .dispatch_request(*args, **kwargs)
  File "flask/views.py", line 149, in dispatch_request
    return meth(*args, **kwargs)
  File "invenio_records_rest/views.py", line 424, in get
    search_result = search.execute()
  File "elasticsearch_dsl/search.py", line 579, in execute
    **self._params
  File "elasticsearch/client/utils.py", line 69, in _wrapped
    return func(*args, params=params, **kwargs)
  File "elasticsearch/client/__init__.py", line 548, in search
    doc_type, '_search'), params=params, body=body)
  File "elasticsearch/transport.py", line 329, in perform_request
    status, headers, data = connection.perform_request(method, url, params, body, ignore=ignore, timeout=timeout)
  File "elasticsearch/connection/http_requests.py", line 84, in perform_request
    self._raise_error(response.status_code, raw_data)
  File "elasticsearch/connection/base.py", line 108, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
```

### PR DESCRIPTION
**Pull request by JavierDelgadoFernandez**
* Adds a handler that process the errors coming from ElasticSearch in
  case of a bad request. (closes #84)

* Fixes issue with `Sphinx 1.4`.

Signed-off-by: Javier Delgado <javier.delgado.fernandez@cern.ch>

### CODE REVIEW

File: invenio_records_rest/views.py [hunk 1]
```diff
@@ -74,6 +76,11 @@ def validation_error(error):
         """Catch validation errors."""
         return RESTValidationError().get_response()
 
+    @blueprint.errorhandler(RequestError)
+    def elasticsearch_badrequest_error(error):
+        """Catche problems in ElasticSearch queries."""
+        return BadRequest(description=error.error).get_response()
```
Comment (lnielsen) [inline invenio_records_rest/views.py]: The error message has to be rewritten to something meaningful for an end-user as the message is display upfront in the UI. Please also see ``invenio_rest.errors``

Comment (JavierDelgadoFernandez) [inline invenio_records_rest/views.py]: I checked the exceptions in ``invenio_rest.errors`` but there is no one that fits properly here. 

About the description, what you are getting in this case is a ``search_phase_execution_exception``. Maybe not in all cases the error is going to be the same. I can rewrite the error description, ~~but isn't better to use ``{% trans %}`` in the view part with the error response?~~Ok, I understand that Angular is making things more complicated here.

I am checking now why Zenodo is not showing the error descriptions. It is just showing ``Error:`` and nothing else
zenodo_error
:

File: invenio_records_rest/views.py [hunk 2]
```diff
@@ -74,6 +76,12 @@ def validation_error(error):
         """Catch validation errors."""
         return RESTValidationError().get_response()
 
+    @blueprint.errorhandler(RequestError)
+    def elasticsearch_badrequest_error(error):
+        """Catche problems in ElasticSearch queries."""
+        description = gettext(error.error)
```
Comment (lnielsen) [inline invenio_records_rest/views.py]: Can you provide some examples of what the text is. This error message is for end-users so it's very important that it's something as simple as "The syntax of the search query is invalid."

Comment (JavierDelgadoFernandez) [inline invenio_records_rest/views.py]: @lnielsen Ok, I am not using the exception message then. It looks like this right now:
zenodo_error2

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen Updated. Here is the JSON coming from ES:
```json
{

   "status": 400,

   "error": {

      "phase": "query",

      "type": "search_phase_execution_exception",

      "root_cause": [

         {

            "index": "invenio-records-rest-testrecord",

            "col": 144,

            "reason": "Failed to parse query [test:]",

            "line": 1,

            "type": "query_parsing_exception"

         }

      ],

      "failed_shards": [

         {

            "index": "invenio-records-rest-testrecord",

            "node": "4renZaHOTC-9iKqDQODXfg",

            "shard": 0,

            "reason": {

               "line": 1,

               "type": "query_parsing_exception",

               "index": "invenio-records-rest-testrecord",

               "col": 144,

               "reason": "Failed to parse query [test:]",

               "caused_by": {

                  "reason": "Cannot parse 'test:': Encountered \"<EOF>\" at line 1, column 5.\nWas expecting one of:\n    <BAREOPER> ...\n    \"(\" ...\n    \"*\" ...\n    <QUOTED> ...\n    <TERM> ...\n    <PREFIXTERM> ...\n    <WILDTERM> ...\n    <REGEXPTERM> ...\n    \"[\" ...\n    \"{\" ...\n    <NUMBER> ...\n    ",

                  "caused_by": {

                     "reason": "Encountered \"<EOF>\" at line 1, column 5.\nWas expecting one of:\n    <BAREOPER> ...\n    \"(\" ...\n    \"*\" ...\n    <QUOTED> ...\n    <TERM> ...\n    <PREFIXTERM> ...\n    <WILDTERM> ...\n    <REGEXPTERM> ...\n    \"[\" ...\n    \"{\" ...\n    <NUMBER> ...\n    ",

                     "type": "parse_exception"

                  },

                  "type": "parse_exception"

               }

            }

         }

      ],

      "reason": "all shards failed",

      "grouped": true

   }

}
```

File: invenio_records_rest/views.py [hunk 3]
```diff
@@ -56,6 +58,18 @@
     lambda: current_app.extensions['invenio-records-rest'])
 
 
+def elasticsearch_query_parsing_exception_handler(error):
+    """Handle query parsing exceptions from ElasticSearch."""
+    description = gettext('The syntax of the search query is invalid.')
```
Comment (jirikuncar) [inline invenio_records_rest/views.py]: for consistency reasons can you import `gettext` as `_`?

File: invenio_records_rest/views.py [hunk 4]
```diff
@@ -30,9 +30,11 @@
 import uuid
 from functools import partial, wraps
 
+from elasticsearch.exceptions import RequestError
 from flask import Blueprint, abort, current_app, jsonify, make_response, \
     request, url_for
 from flask.views import MethodView
+from flask_babelex import gettext
```
Comment (jirikuncar) [inline invenio_records_rest/views.py]: - [ ] please include `Flask-BabelEx` in `setup.py`
- [ ] add configuration for `Babel` to `setup.cfg`
- [ ] run `python setup.py extract_messages` and initialize catalog for english
- [ ] add `.tx` config

(you can generate most of it using cookiecutter-invenio-module)

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar @lnielsen Updated.

**Comment by jirikuncar** [comment]
@JavierDelgadoFernandez This branch is out-of-date with the base branch.

File: .tx/config [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+
+[main]
+host = https://www.transifex.com
+
+[invenio-records-rest.messagespot]
+file_filter = invenio_records_rest/translations/<lang>/LC_MESSAGES/messages.po
+source_file = invenio_records_rest/translations/messages.pot
+source_lang = en
+type = PO
```
Comment (jirikuncar) [inline .tx/config]: please end file with `\n`.

File: babel.ini [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+# Extraction from Python source files
+
+[python: **.py]
+encoding = utf-8
```
Comment (jirikuncar) [inline babel.ini]: https://github.com/inveniosoftware/invenio-records-rest/pull/87/files#r65887590

File: setup.cfg [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 #
 # This file is part of Invenio.
-# Copyright (C) 2015, 2016 CERN.
+# Copyright (C) 2016 CERN.
```
Comment (jirikuncar) [inline setup.cfg]: please keep the original years

File: setup.cfg [hunk 8]
```diff
@@ -22,8 +22,6 @@
 # waive the privileges and immunities granted to it by virtue of its status
 # as an Intergovernmental Organization or submit itself to any jurisdiction.
 
-[aliases]
```
Comment (jirikuncar) [inline setup.cfg]: it is needed

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen @jirikuncar Updated.

**Pull request merged**